adding interface files and unit tests

gabrielcocenza opened the pull request:
This PR attempts to add a new interface that will support the integration between designate-bind and prometheus-bind-exporter-operator.

After the approval, I will open a PR in [juju layer-index](https://github.com/juju/layer-index) in order to be able to fetch directly when building the `charm-designate-bind`

Closes-Bug: [#1938888](https://bugs.launchpad.net/charm-designate-bind/+bug/1938888)
[Gerrit-PR](https://github.com/canonical/interface-bind-client/compare/master...gabrielcocenza:bug/1938888?expand=1)

gabrielcocenza committed:
adding interface files and unit tests

auria commented on provides.py:
```diff
@@ -0,0 +1,53 @@
+
+    @reactive.when("endpoint.{endpoint_name}.departed")
+    def departed(self):
+        self.remove()
```
I checked what [`clear_flags`](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/flags.py#L95) and [`is_flag_set`](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/flags.py#L206) do, and I think the state that should be created is `{endpoint_name}.connected`. This would allow to react when a relation exists or has departed, and the broken/departed hooks would always assume that the state exists.

Charms could react to the `.connected` state by creating another state that could vary (be enabled or not), depending on configurations being up to date or changes being triggered.

I find it tricky that a state create by this interface can be later removed by the charm.

If the approach described here is taken, I think the last 3 methods (remove, broken, departed) could be merged into a single one:
```
@reactive.when_any("endpoint.{endpoint_name}.broken", "endpoint.{endpoint_name}.departed")
def remove(self):
    reactive.clear_flag(self.expand_name("{endpoint_name}.connected"))
```

gabrielcocenza replied:
I've changed the flag to `connected`, but as I said in the other comment, unfortunately the framework doesn't pass argument to the handler when using `when_any`

auria commented on requires.py:
```diff
@@ -0,0 +1,35 @@
+    def joined(self):
+        reactive.set_flag(self.expand_name('{endpoint_name}.available'))
+
+    def remove(self):
+        reactive.clear_flag(self.expand_name('{endpoint_name}.available'))
+
+    @reactive.when('endpoint.{endpoint_name}.broken')
+    def broken(self):
+        self.remove()
+
+    @reactive.when('endpoint.{endpoint_name}.departed')
+    def departed(self):
+        self.remove()
```
For alignment with the provides function, this would be changed to use a ".connected" state (if the approach described for the provides interface makes sense).

In any case, the last 3 methods (remove, broken, departed) could be merged into a single one as mentioned above (using `reactive.when_any`).

rgildein replied:
```suggestion
    @reactive.when_any('endpoint.{endpoint_name}.broken', 'endpoint.{endpoint_name}.departed')
    def broken(self):
        reactive.clear_flag(self.expand_name('{endpoint_name}.available'))
```

rgildein commented on interface.yaml:
```diff
@@ -0,0 +1,12 @@
+summary: Interface for intergrating bind
+maintainer: DevOps Centres
+ignore:
+  - 'unit_tests'
```
nitpick: are quotes needed?

gabrielcocenza replied:
I'm not sure if the quotes are needed, but all the examples that I found are using it.

gabrielcocenza replied:
I've tried to remove the quotes and it seems that works normally. Will remove it.

rgildein commented on provides.py:
```diff
@@ -0,0 +1,53 @@
+            reactive.set_flag(self.expand_name("{endpoint_name}.available"))
+            self.set_ingress_address()
+
+    def remove(self):
+        if self.available():
+            reactive.clear_flag(self.expand_name("{endpoint_name}.available"))
+
+    @reactive.when("endpoint.{endpoint_name}.broken")
+    def broken(self):
+        self.remove()
+
+    @reactive.when("endpoint.{endpoint_name}.departed")
+    def departed(self):
+        self.remove()
```
```suggestion
    @reactive.when_any("endpoint.{endpoint_name}.broken", "endpoint.{endpoint_name}.departed")
    def broken(self):
        if self.available():
            reactive.clear_flag(self.expand_name("{endpoint_name}.available"))
```

gabrielcocenza replied:
@auria and @r-f-g The problem that happens when it uses `when_any` is that the handler never gets the `self` argument as it says in the [charm-reactive](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/decorators.py#L134-L143). I've tried also to make as a classmethod, but it doesn't work also as it shows in this [logs](https://pastebin.canonical.com/p/ssfbF7zY25/)

This part of the [documentation](https://charmsreactive.readthedocs.io/en/latest/charms.reactive.decorators.html#charms-reactive-decorators) also confirms this behavior of not passing arguments.

I think we should keep the way as it is implemented with one flag for each endpoint handler.

rgildein replied:
@gabrielcocenza I don't think that you are right. I look at the [charm-reactive](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/decorators.py#L84) and I understand it that it should works. I think that there will no other arguments, but `self` will be there. Did you tried it and it was not working?

I found some examples:
[charm-interface-cinder-backu](https://opendev.org/openstack/charm-interface-cinder-backup/src/branch/master/provides.py#L35)
[charm-interface-websso-fid-service-provider](https://opendev.org/openstack/charm-interface-websso-fid-service-provider/src/branch/master/provides.py#L37)
[charm-interface-barbican-secrets](https://github.com/openstack/charm-interface-barbican-secrets/blob/master/requires.py#L24)
[charm-interface-placement](https://github.com/openstack/charm-interface-placement/blob/master/provides.py#L35)

gabrielcocenza replied:
@r-f-g  I discovered that the problem was that the framework is not capable of recognize the endpoint if you use `class BindClientProvides(reactive.Endpoint)` instead of `class BindClientProvides(Endpoint)`. I think this is kind of a bug, but in anyway, I've changed and now it works.

Thanks for checking this.

gabrielcocenza replied:
@r-f-g, @auria just found this issue in the [reactive framework](https://github.com/juju-solutions/charms.reactive/issues/184)

rgildein replied:
Thank you for your investigation.:+1:

rgildein reviewed (commented):
Overall, LGTM. I added one nitpick and two suggestions for shorter code.

auria reviewed (changes requested):
The `peers.py` script should be removed since there are no peer relations supported by this interface.

Please find other comments inline. Thank you!

auria commented:
Regarding the README, I think the `stats-listen-net` should not be shared via the relation. That information in Juju config would help the charm determine which IP address is part of the shared network block. In case a machine has more than one interface, this network block would determine that it's eth1 and not eth0 the one to be used for listening.

The applications that receive the data would need a specific IP address to connect to, not a network block.

This interface would allow to configure this relation between 2 non-subordinate charms, so a machine would connect to another machine (Note: in our use case, one of the charms is subordinate, so localhost will be used instead, but it may not always be the case).

On the other hand, if my sugestion about using `{endpoint_name}.connected` is accepted, that info would need to be updated in the README, too.

Thank you!

gabrielcocenza committed:
changed flag from available to connected and fixed README

gabrielcocenza commented:
Thanks for the review guys.

I've updated the flag to use `connected`  instead of `available` and updated the README.

In one of the comments above, I pointed out why I think we should not merge the endpoints flags using `when_any`. Basically is because this decorator doesn't pass the argument `self` (Endpoint object).

Regards

auria commented on unit_tests/test_requires.py:
```diff
@@ -0,0 +1,81 @@
+# limitations under the License.
+
+import charms_openstack.test_utils as test_utils
+import mock
```
You may want to replace `import mock` by `from unittest.mock as mock`. By doing so, you will not need to install the "mock" package via `test-requirements.txt`.

gabrielcocenza replied:
The mock is a dependence of the module `test_utils` in charms_openstack. I've left a comment in the `test-requirements.txt`

auria commented on unit_tests/test_provides.py:
```diff
@@ -0,0 +1,81 @@
+# limitations under the License.
+
+import charms_openstack.test_utils as test_utils
+import mock
```
You may want to replace `import mock` by `from unittest.mock as mock`. By doing so, you will not need to install the "mock" package via `test-requirements.txt`.

auria commented on requires.py:
```diff
@@ -0,0 +1,32 @@
+
+    @reactive.when("endpoint.{endpoint_name}.broken")
+    def broken(self):
+        reactive.clear_flag(self.expand_name("{endpoint_name}.connected"))
```
Same comment as previous time. Can't `departed` and `broken` decorated methods be merged into a single method and use `reactive.when_any` ?

auria commented on provides.py:
```diff
@@ -0,0 +1,32 @@
+
+    @reactive.when("endpoint.{endpoint_name}.broken")
+    def broken(self):
+        reactive.clear_flag(self.expand_name("{endpoint_name}.connected"))
```
Same comment as previous time. Can't `departed` and `broken` decorated methods be merged into a single method and use `reactive.when_any` ?

auria commented on requires.py:
```diff
@@ -0,0 +1,32 @@
+from charms import reactive
+
+
+class BindClientRequires(reactive.Endpoint):
```
This class misses the `__init__` constructor, inheriting from `Endpoint`. The provider class does define the constructor. Is this intended?

OTOH, I think this method could benefit from a method that collects possible attributes shared over the relation (e.g. `hostname` and `port`). This approach may not immediately help the `prometheus-bind-exporter-operator` charm because it is not reactive. But if there's ever a reactive charm using this interface, it won't need to use `relation_get` directly.

gabrielcocenza replied:
>The provider class does define the constructor. Is this intended?

It wasn't intended. Great catch. Thanks!

>But if there's ever a reactive charm using this interface, it won't need to use relation_get directly.

I've searched a method in the `Endpoint` object and didn't find a method for this. The equivalent of relation_set can be used in the `provides.py`  with `publish_raw`. Do you have an example of this?

gabrielcocenza replied:
I thought doing something like this:
```python
    def get_config(self):
        """
        Returns a dict of available designate-bind services and their associated IP
        and ports.
        The return value is a dict of the following form::
            {
                'unit_name': {
                    'port': port_of_designate_bind_unit,
                    'ip': ip_of_designate_bind_unit
                },
                # ...
            }
        """
        configs = {}
        for relation in self.relations:
            for unit in relation.joined_units:
                unit_name = relation.unit_name
                config = configs.setdefault(unit_name, {
                    unit_name: {}
                })
                data = unit.received_raw
                port = data.get('port')
                ip = data.get('ip')
                if port and ip:
                    config[unit_name] = {'port': port, 'ip': ip}

        return configs
```

auria commented on provides.py:
```diff
@@ -0,0 +1,32 @@
+from charms import reactive
+
+
+class BindClientProvides(reactive.Endpoint):
```
I wonder if this interface should implement a method to set attributes (passed via arg) to be shared over the relation. For example, the [`http` interface](https://github.com/juju-solutions/interface-http/blob/master/provides.py#L25) implements a `configure` method.

gabrielcocenza replied:
I think it's a nice idea and I implemented to pass the port with a `configure` method. I've also added in the README how to use it.

auria reviewed (commented):
I've let some more comments inline. Thank you.

gabrielcocenza committed:
new method to configure port, fix CR comments

auria commented on interface.yaml:
```diff
@@ -0,0 +1,12 @@
+name: bind-client
+summary: Interface for intergrating bind
+maintainer: DevOps Centres
```
This should have an email address. You may want to put your name and canonical's email address since the DevOps Centre does not have one.

auria commented on README.md:
```diff
@@ -20,25 +20,29 @@ requires:
+    endpoints = [
+        reactive.endpoint_from_flag('dns-backend.related'),
+        reactive.endpoint_from_name('bind-stats')
+    ]
```
This code needs to specify that for the example, the `metadata.yaml` file of this charm would have a "dns-backend" binding and a "bind-stats" endpoint using the `bind-client` interface. Otherwise, it is not clear where the "dns-backend.related" flag or the "bind-stats" endpoint come from. It is also unclear where "designate_bind" object comes from (lines 44-45).

OTOH, I would split this list in separate variables (see example below). This is because calling endpoints[1] is not immediately understandable (line 42).
```
dns_backend_endpoint = reactive.endpoint_from_flag("dns-backend.related")
bind_stats_endpoint = reactive.endpoint_from_name("bind-stats")
config = hookenv.config()

if reactive_is_flag_set("bind-stats.connected") and config.changed("stats-port"):
    bind_stats_endpoint.configure(config["stats-port"])

# etc.
```

auria commented on requires.py:
```diff
@@ -0,0 +1,41 @@
+
+    @reactive.when("endpoint.{endpoint_name}.broken")
+    def broken(self):
+        reactive.clear_flag(self.expand_name("{endpoint_name}.connected"))
```
I think this code misses a way to retrieve the "port" attribute shared over the relation?

gabrielcocenza replied:
Wouldn't this be at provides instead of requires? I didn't see this kind of logic in another relations, but I think it makes sense to have it.

How do you think this should be done?

Checking charm-helpers it has the function [relation_clear](https://github.com/juju/charm-helpers/blob/master/charmhelpers/core/hookenv.py#L532-L540) that is passing `None` to values.

I though doing something like this:
```python
    @reactive.when("endpoint.{endpoint_name}.broken")
    def broken(self):
        reactive.clear_flag(self.expand_name("{endpoint_name}.connected"))
        for relation in self.relations:
            relation.to_publish_raw['port'] = None
            relation.to_publish_raw['ip'] = None
```

rgildein replied:
This is quite interesting, if you set something `None` it will be not present in relation data? It thinks it's better to erase it, don't let it be there.

gabrielcocenza replied:
Yes, passing `None` `Juju` understands that it should not show those parameters.  I think we can keep this way and we can use the same method to pass values or erase it.

What do you think?

auria reviewed (changes requested):
Added some comments inline.

rgildein reviewed (commented):
I left two comments.

gabrielcocenza committed:
update readme, added configure in provides and added get_config in requires

auria commented on provides.py:
```diff
@@ -28,22 +28,21 @@ def __init__(self, *args, **kwargs):
 
-    def configure(self, port):
-        """Configure the port to be used in the relation between units
+    def configure(self, shared_data):
```
By using a dict (`shared_data`), you are essentially giving total freedom to the layers that use this interface to set any key-values they want. Is this intended? Shouldn't a list of args be shared to make it explicit what attributes are expected? Note that having an attribute's value set to `None` would be that the attribute would not be shared (e.g. `hostname`, because the default is to use localhost).

gabrielcocenza replied:
The original idea it was to give total freedom to share any relevant data in the relation, but as you pointed it might not be a good idea, but I think it's a good thing that we can pass `None` as parameters to erase port and IP when the relation is `departed` or `broken`.

auria commented on README.md:
```diff
@@ -4,53 +4,92 @@ This interface supports integration between designate-bind and prometheus-bind-e
-            endpoints[1].configure(hookenv.config('stats-port'))
+            shared_data = {
+                'stats-port': hookenv.config('stats-port'),
+                'stats-ip': '127.0.0.1'
```
It may be redundant to use the `stats-` prefix. What do you think?

gabrielcocenza replied:
I agree that it's redundant

auria reviewed (commented):
I've added a couple of comments which don't block the approval, though.

gabrielcocenza committed:
config method more restrict, merge departed and broken flags

rgildein reviewed (approved):
LGTM

auria commented on requires.py:
```diff
@@ -0,0 +1,73 @@
+                if port and ip:
+                    configs[unit_name] = {'port': port, 'ip': ip}
+
+        return configs
```
I think I pointed you towards the wrong example (http interface). Sorry for that.

Since this interface is meant to work between a principle charm and a subordinate charm, the scope of this method should be a single unit. So the config returned should be a dict with the key-values shared by the principle unit (e.g. designate-bind).

You can find a better example implemented in charm-telegraf:
* [layer.yaml](https://git.launchpad.net/charm-telegraf/tree/src/layer.yaml) lists the [elasticsearch interface](https://github.com/juju-solutions/interface-elasticsearch/blob/master/requires.py)
* the [metadata.yaml](https://git.launchpad.net/charm-telegraf/tree/src/metadata.yaml#n45) lists the `elasticsearch` interface with container scope. So, the `requires.py` from the interface shows `scope = scopes.UNIT`. However, that class inherits from `RelationBase`, not from `Endpoint`, so I'm not sure if the behavior would be the same. The point is that you should only parse an return the info from the principle unit your subordinate unit depends (if it makes sense). You don't want to retrieve the info from all the principle units because the scope should be limited.

gabrielcocenza replied:
Reading the [reactive Endpoint](https://github.com/juju-solutions/charms.reactive/blob/f4f5ac61713a5544c3c522fa5873157a383241d7/charms/reactive/endpoints.py) it seems that the only way to get access to relate units is through [relations](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/endpoints.py#75).

If I understand you correctly, your concern is when, for example, two units of `charm-designate-bind` are in the deployed bundle:
```
Unit                Workload  Agent      Machine  Public address  Ports     Message
designate-bind/0*   active    executing  0        10.5.2.54                 (resume) Unit is ready
  bind-exporter/0*  active    idle                10.5.2.54                 Unit is ready
designate-bind/1    active    idle       1        10.5.3.211                Unit is ready
  bind-exporter/1   active    idle                10.5.3.211                Unit is ready
```
And this would result into something like this when using the method `get_config`
```
{
   "designate-bind/0":{
      "port":"8053",
      "ip":"127.0.0.1"
   },
   "designate-bind/1":{
      "port":"8053",
      "ip":"127.0.0.1"
   }
}
```
To say the truth, that was my idea, but It turns out that this doesn't happen. Using `pdb` to this bundle above, I could see that the Endpoint will just have one `endpoint.Relation` object which will have just one `endpoints.RelatedUnit` (`joined_units`) that is directly connected. In another words, it seems to me that the `Endpoint`  uses by default a `Unit scope` and won't get data from another units.

I think we can pass directly a Dict containing the `port` and `ip` without adding the unit name, but the way to get it using two nested for loops I think it's correct.

What do you think?

auria replied:
Per conversation offline, the next step is to clarify if the Endpoint class has a unit scope by default (or only when the metadata.yaml file determines that an endpoint has `scope: container`). We don't want to have a method that returns the last dict formed from the relation data of the last unit parsed (we want to be 100% sure only the data from a single unit [the principle] is accessible by the subordinate unit (no matter other units+subordinates may exist).

auria replied:
Let's keep it as is and then use `.values()[0]` on subordinate charms. I think that approach will allow the requires implementation to be used on non-subordinate charms, too (e.g. a single unit that configures connections against each of the providers)

gabrielcocenza replied:
Talking with Cory was clarifying about this topic.
> RelationBase tried to work was to define ahead of time whether it would be handling every single unit of every related app individually, treating each related app as a single unit, or just publishing something globally and not caring about who was listening.  That's what the scope was. It turned out that this approach was very confusing and since it didn't map well to the way relations actually worked, it made it really hard to figure out how to write a relation.

> Essentially, each relation is a collection of key-value buckets, one for each app and one for each unit of each app.  A given unit can set data in it's unit data bucket, and see the data in the bucket for each other unit on the relation. Only leader units can read or write the app level data buckets.  Everything in the Python is just an attempt to present that underlying reality to the charm.

> But subordinate charms are special cases, as I said, just in which units are visible to each unit on the relation.  You can kind of think of a unit of a subordinate as it's own individual application with respect to the principal charm it's attached to (although they act like normal apps if you have other, "normal" relations to other charms).

So basically, the relation is having this behavior because it's subordinate and if wasn't, it would be able to read the information from other units. As we said in conversation offline, the current approach could fit for both cases (subordinate and not). In case the charm is subordinate it could just access the first value in the dict.

auria reviewed (changes requested):
I've added a comment inline. Sorry I didn't catch it earlier.

auria reviewed (approved):
FYI, I can't resolve conversations within the PR but I'm +1 to merge your change.
